File: backend_node/src/graph/state.ts
import { Annotation } from "@langchain/langgraph";

export const GraphState = Annotation.Root({
    question: Annotation<string>,
    schema: Annotation<string>,
    db: Annotation<any>,
    sql: Annotation<string>,
    valid: Annotation<boolean>,
    feedback: Annotation<string>,
    result: Annotation<any>,
    iterations: Annotation<number>,
    restrictedColumns: Annotation<string[]>,
});

File: backend/app/graph/state.py
from typing import TypedDict, Annotated, List, Any, Optional
from langchain_core.messages import BaseMessage
import operator

class GraphState(TypedDict):
    question: str
    schema: str
    sql: Optional[str]
    result: Optional[Any]
    feedback: Optional[str]
    valid: bool
    iterations: int
    db_path: Optional[str]
    restricted_columns: Optional[List[str]]
    intent: Optional[str]
    upload_id: Optional[str]
